tools/IoTEdgeDevOps/DevOpsLib/AzureActiveDirectory.cs

```diff
@@ -0,0 +1,93 @@
+        {
+            ValidationUtil.ThrowIfNullOrWhiteSpace(azureActiveDirTenant, nameof(azureActiveDirTenant));
+            ValidationUtil.ThrowIfNullOrWhiteSpace(azureActiveDirClientId, nameof(azureActiveDirClientId));
+            ValidationUtil.ThrowIfNullOrWhiteSpace(azureActiveDirClientSecret, nameof(azureActiveDirClientSecret));
+            ValidationUtil.ThrowIfNullOrWhiteSpace(azureResource, nameof(azureResource));
+
+            try
+            {
+                
+                if (this.IsAccessTokenExpired() ||
+                    (this.accessToken == null) ||
+                    (!this.azureResource.Equals(azureResource, StringComparison.OrdinalIgnoreCase)))
+                {
+                    string requestUri = $"https://login.microsoftonline.com/{azureActiveDirTenant}/oauth2/token";
+                    const string grantType = "client_credentials";
+
+                    var client = new HttpClient();
+                    client.BaseAddress = new Uri(requestUri);
+
+                    var requestBody = new List<KeyValuePair<string, string>>();
+                    requestBody.Add(new KeyValuePair<string, string>("client_id", azureActiveDirClientId));
+                    requestBody.Add(new KeyValuePair<string, string>("client_secret", azureActiveDirClientSecret));
+                    requestBody.Add(new KeyValuePair<string, string>("grant_type", grantType));
+                    requestBody.Add(new KeyValuePair<string, string>("resource", azureResource));
+
+                    HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Post, "");
+                    // By default, if FormUrlEncodedContent() is used, the "Content-Type" is set to "application/x-www-form-urlencoded"
+                    // which can be explicitly written : request.Content.Headers = new MediaTypeHeaderValue("application/x-www-form-urlencoded");
+                    request.Content = new FormUrlEncodedContent(requestBody);
+
+                    HttpResponseMessage response = await client.SendAsync(request).ConfigureAwait(false);
+                    response.EnsureSuccessStatusCode();
+                    string responseMsg = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+
+                    lock (locker)
+                    {
+                        var responseJson = JObject.Parse(responseMsg);
+                        this.azureResource = azureResource;
+                        this.accessTokenExpiration = DateTime.UtcNow.AddSeconds((double)responseJson["expires_on"] - 1);
+                        this.accessToken = (string)responseJson["access_token"];
+                    }
+                }
+
+                return this.accessToken;
+            }
+            catch (Exception e)
+            {
+                throw e;
+            }
+        }
+
+        public bool IsAccessTokenExpired()
+        {
+            return DateTime.Compare(DateTime.UtcNow, this.accessTokenExpiration) >= 0;
```

Comment:
Add buffer to this checking.

Reply:
I think it is easier to read (with buffer, you can make buffer timespan as a readonly variable).

e.g. (((DateTime.UtcNow.AddDays(3) - DateTime.UtcNow) - new TimeSpan(0,0,0,5)).Ticks > 0)